File: frontend/src/api/medicationReports.ts
import { apiRequest } from './client'
import type { MedicationCheckinStatus } from './medicationCheckins'
import type { MedicationOccurrenceStatus } from './medicationOccurrences'
import type { BarrierCode, SubreasonCode, SupportCode } from './trackC'

// Contract sources:
// - backend/app/apis/v1/medication_report_routers.py
// - backend/app/dtos/medication_reports.py

export type MedicationReportPeriod = 7 | 30
export type MedicationReportTimeSlot = 'BREAKFAST' | 'LUNCH' | 'DINNER' | 'BEDTIME'

export type MedicationReportCounts = {
  taken_count: number
  not_taken_count: number
  unconfirmed_count: number
  pending_count: number
  cancelled_count: number
}

export type MedicationReportRate = {
  numerator: number
  denominator: number
  percentage: number | null
}

export type MedicationReportRecord = {
  occurrence_id: string
  prescription_version_id: string
  prescription_version_medication_id: string
  scheduled_local_date: string
  scheduled_at: string
  confirmation_deadline_at: string
  status: MedicationOccurrenceStatus
  medication_name: string
  strength_text: string | null
  time_slot: MedicationReportTimeSlot
  updated_at: string
  checkin: {
    checkin_id: string
    occurrence_id: string
    status: MedicationCheckinStatus
    taken_at: string | null
    revision: number
    corrected: boolean
    updated_at: string
  } | null
}

// Missed-dose reasons and consultation questions are returned only for the clinic
// view; the default report never carries them.
export type MedicationReportView = 'CLINIC'

export type ClinicBarrierEntry = {
  occurrence_id: string
  scheduled_local_date: string
  medication_name: string
  barrier_code: BarrierCode
  subreason_code: SubreasonCode | null
}

export type ClinicConsultationQuestion = {
  question_id: string
  text: string
  support_code: SupportCode
  medication_name: string
  last_selected_date: string
}

export type ClinicSections = {
  barriers: ClinicBarrierEntry[]
  consultation_questions: ClinicConsultationQuestion[]
}

export type MedicationReportData = {
  period_days: MedicationReportPeriod
  start_date: string
  end_date: string
  timezone: 'Asia/Seoul'
  as_of: string
  counts: MedicationReportCounts
  overdue_pending_count: number
  adherence_rate: MedicationReportRate
  confirmation_rate: MedicationReportRate
  records: MedicationReportRecord[]
  clinic: ClinicSections | null
}

export type MedicationReportResponse = {
  data: MedicationReportData
}

export async function getMedicationReport(
  periodDays: MedicationReportPeriod,
  signal?: AbortSignal,
  view?: MedicationReportView,
): Promise<MedicationReportResponse> {
  const viewQuery = view ? `&view=${view}` : ''

  return apiRequest<MedicationReportResponse>(
    `/api/v1/medication-reports?period_days=${periodDays}${viewQuery}`,
    { signal },
  )
}

File: backend/app/dtos/medication_reports.py
from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, Field

from app.dtos.medication_checkins import MedicationCheckinData
from app.dtos.medication_schedules import MedicationOccurrenceData
from app.models.track_c import BarrierCode, SupportCode

# The clinic view is the only surface approved to show why a dose was missed.
MedicationReportView = Literal["CLINIC"]


class MedicationReportCounts(BaseModel):
    taken_count: int = Field(default=0, ge=0)
    not_taken_count: int = Field(default=0, ge=0)
    unconfirmed_count: int = Field(default=0, ge=0)
    pending_count: int = Field(default=0, ge=0)
    cancelled_count: int = Field(default=0, ge=0)


class MedicationReportRate(BaseModel):
    numerator: int = Field(ge=0)
    denominator: int = Field(ge=0)
    percentage: float | None = Field(ge=0, le=100)


MedicationReportTimeSlot = Literal["BREAKFAST", "LUNCH", "DINNER", "BEDTIME"]


class MedicationReportCheckin(MedicationCheckinData):
    updated_at: AwareDatetime


class MedicationReportRecord(MedicationOccurrenceData):
    medication_name: str
    strength_text: str | None
    time_slot: MedicationReportTimeSlot
    checkin: MedicationReportCheckin | None
    updated_at: AwareDatetime


class ClinicBarrierEntry(BaseModel):
    occurrence_id: UUID
    scheduled_local_date: date
    medication_name: str
    barrier_code: BarrierCode
    subreason_code: str | None = None


class ClinicConsultationQuestion(BaseModel):
    question_id: str
    text: str
    support_code: SupportCode
    medication_name: str
    last_selected_date: date


class ClinicSections(BaseModel):
    """Only populated for view=CLINIC; absent everywhere else by construction."""

    barriers: list[ClinicBarrierEntry]
    consultation_questions: list[ClinicConsultationQuestion]


class MedicationReportData(BaseModel):
    period_days: Literal[7, 30]
    start_date: date
    end_date: date
    timezone: Literal["Asia/Seoul"] = "Asia/Seoul"
    as_of: AwareDatetime
    counts: MedicationReportCounts
    overdue_pending_count: int = Field(ge=0)
    adherence_rate: MedicationReportRate
    confirmation_rate: MedicationReportRate
    records: list[MedicationReportRecord]
    clinic: ClinicSections | None = None


class MedicationReportResponse(BaseModel):
    data: MedicationReportData
